src/script/tari_script.rs

```diff
@@ -309,50 +500,310 @@ impl fmt::Display for TariScript {
+    if_count: u16,
+    else_count: u16,
+}
 
-impl Builder {
-    pub fn new() -> Self {
-        Builder {}
+impl Default for ScriptState {
+    fn default() -> Self {
+        Self {
+            executing: true,
+            if_count: 0,
+            else_count: 0,
+        }
     }
 }
 
 #[cfg(test)]
 mod test {
+    #[allow(deprecated)]
     use crate::{
         common::Blake256,
+        hash::sha3::Sha3,
         inputs,
         keys::{PublicKey, SecretKey},
         ristretto::{pedersen::PedersenCommitment, RistrettoPublicKey, RistrettoSchnorr, RistrettoSecretKey},
-        script::{error::ScriptError, op_codes::to_boxed_hash, ExecutionStack, TariScript, DEFAULT_SCRIPT_HASH},
+        script::{
+            error::ScriptError,
+            op_codes::{to_boxed_hash, to_boxed_message},
+            ExecutionStack,
+            ScriptContext,
+            StackItem::{Commitment, Hash, Number},
+            TariScript,
+            DEFAULT_SCRIPT_HASH,
+        },
     };
     use blake2::Digest;
+    use sha2::Sha256;
     use tari_utilities::{hex::Hex, ByteArray};
 
     #[test]
     fn default_script() {
-        let script = TariScript::default();
+        let mut script = TariScript::default();
         let inputs = ExecutionStack::default();
         assert!(script.execute(&inputs).is_ok());
         assert_eq!(&script.to_hex(), "7b");
         assert_eq!(script.as_hash::<Blake256>().unwrap(), DEFAULT_SCRIPT_HASH);
     }
 
+    #[test]
+    fn op_or() {
+        let mut script = script!(Or(1));
+
+        let inputs = inputs!(4, 4);
+        let result = script.execute(&inputs).unwrap();
+        assert_eq!(result, Number(1));
+
+        let inputs = inputs!(3, 4);
+        let result = script.execute(&inputs).unwrap();
+        assert_eq!(result, Number(0));
+
+        let mut script = script!(Or(3));
+
+        let inputs = inputs!(1, 2, 1, 3);
+        let result = script.execute(&inputs).unwrap();
+        assert_eq!(result, Number(1));
+
+        let inputs = inputs!(1, 2, 4, 3);
+        let result = script.execute(&inputs).unwrap();
+        assert_eq!(result, Number(0));
+
+        let inputs = inputs!(2, 4, 3);
+        let err = script.execute(&inputs).unwrap_err();
+        assert!(matches!(err, ScriptError::StackUnderflow));
+
+        let mut script = script!(OrVerify(1));
+
+        let inputs = inputs!(1, 4, 4);
+        let result = script.execute(&inputs).unwrap();
+        assert_eq!(result, Number(1));
+
+        let inputs = inputs!(1, 3, 4);
+        let err = script.execute(&inputs).unwrap_err();
+        assert!(matches!(err, ScriptError::VerifyFailed));
+
+        let mut script = script!(OrVerify(2));
+
+        let inputs = inputs!(1, 2, 2, 3);
+        let result = script.execute(&inputs).unwrap();
+        assert_eq!(result, Number(1));
+
+        let inputs = inputs!(1, 2, 3, 4);
+        let err = script.execute(&inputs).unwrap_err();
+        assert!(matches!(err, ScriptError::VerifyFailed));
```

Comment:
Try some cases with mixed types, Like a 

Stack:
```
pubkey(Pa)
Number(5)
Pubkey(Pa)
```

`Or(2)` executes, as it does for 

```
pubkey(Pa)
Number(5)
Number(5)
```

Reply:
This wasn't specified - do all `n + 1` items have to be of the same `StackItem` type?

Reply:
Done